### PR TITLE
[NPU] Support repetition penalty for simple generate, Python (cpp backend)

### PR DESCRIPTION
**Pull request by Oscilloscope98**
## Description

https://github.com/analytics-zoo/nano/issues/1741#issuecomment-2531118620

Work with: https://github.com/intel-analytics/llm.cpp/pull/714

- Support repetition penalty for simple generate, Python (cpp backend)
- Fix bug when getting generation config

### CODE REVIEW

File: python/llm/src/ipex_llm/transformers/npu_models/npu_llm_cpp.py [hunk 1]
```diff
@@ -73,16 +78,26 @@ def load_model_from_file(model_dir: str):
...
+                                      repetition_penalty)
     new_token = _lib.llm_sample_token(plogits, True, vocab_size)
     return new_token
 
 
-def run_decode(model_ptr, input_id, vocab_size):
+def run_decode(model_ptr, input_id, vocab_size, updated_input_ids, repetition_penalty=1.0):
     plogits = _lib.run_decode(model_ptr, input_id)
+    updated_input_ptr = (ctypes.c_int32 * len(updated_input_ids))(*updated_input_ids)
+    updated_input_len = len(updated_input_ids)
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_models/npu_llm_cpp.py:96]: Only need these if repetition_penalty != 1

File: python/llm/src/ipex_llm/transformers/npu_models/convert.py [hunk 2]
```diff
@@ -388,27 +387,38 @@ def simple_generate(
...
         eos = [eos]
 
-    output_tokens = []
+    if "repetition_penalty" not in new_generate_kwargs:
+        if hasattr(self.generation_config, "repetition_penalty"):
+            repetition_penalty = self.generation_config.repetition_penalty
+        else:
+            repetition_penalty = 1
+    else:
+        repetition_penalty = new_generate_kwargs['repetition_penalty']
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_models/convert.py:396]:
```python
if "repetition_penalty" in new_generate_kwargs:
   repetition_penalty = new_generate_kwargs['repetition_penalty']
elif hasattr(self.generation_config, "repetition_penalty"):
   repetition_penalty = self.generation_config.repetition_penalty
else:
   repetition_penalty = 1
```

Comment (Oscilloscope98) [reply python/llm/src/ipex_llm/transformers/npu_models/convert.py]: Thank you. Have updated :)

**Review by plusbang — APPROVED** [review]
LGTM

File: python/llm/src/ipex_llm/transformers/npu_models/npu_llm_cpp.py [hunk 3]
```diff
@@ -73,16 +78,26 @@ def load_model_from_file(model_dir: str):
     return _lib.load_model_from_file(model_dir.encode('utf-8'))
 
 
-def run_prefill(model_ptr, input_ids, vocab_size):
+def run_prefill(model_ptr, input_ids, vocab_size, repetition_penalty=1.0):
     input_ptr = (ctypes.c_int32 * len(input_ids))(*input_ids)
     input_len = len(input_ids)
     plogits = _lib.run_prefill(model_ptr, input_ptr, input_len)
+    if repetition_penalty != 1:
+        plogits = _lib.process_logits(plogits, vocab_size,
```
Comment (rnwang04) [inline python/llm/src/ipex_llm/transformers/npu_models/npu_llm_cpp.py:86]: It seems vocab_size is not used, maybe we can remove it.

**Review by rnwang04 — APPROVED** [review]
others LGTM

**Pull request merged**